petab_select/plot.py

```diff
@@ -0,0 +1,777 @@
+        criterion_value = model.get_criterion(criterion)
+        if criterion_value < criterion_value0:
+            criterion_value0 = criterion_value
+            model0 = model
+
+    selected_models = [model0]
+    while True:
+        model0 = selected_models[-1]
+        model1 = model_hashes.get(model0.predecessor_model_hash, None)
+        if model1 is None:
+            break
+        selected_models.append(model1)
+
+    return selected_models[::-1]
+
+
+def get_relative_criterion_values(
+    criterion_values: dict[str, float] | list[float],
+) -> dict[str, float] | list[float]:
+    values = criterion_values
+    if isinstance(criterion_values, dict):
+        values = criterion_values.values()
+
+    value0 = np.inf
+    for value in values:
+        if value < value0:
+            value0 = value
+
+    if isinstance(criterion_values, dict):
+        return {k: v - value0 for k, v in criterion_values.items()}
+    return [v - value0 for v in criterion_values]
+
+
+def upset(
+    models: list[Model], criterion: Criterion
+) -> dict[str, matplotlib.axes.Axes | None]:
+    """Plot an UpSet plot of estimated parameters and criterion.
+
+    Args:
+        models:
+            The models.
+        criterion:
+            The criterion.
+
+    Returns:
+        The plot axes (see documentation from the `upsetplot` package).
+    """
+    # Get delta criterion values
+    values = np.array(
+        get_relative_criterion_values(
+            [model.get_criterion(criterion) for model in models]
+        )
+    )
+
+    # Sort by criterion value
+    index = np.argsort(values)
+    values = values[index]
+    labels = [models[i].get_estimated_parameter_ids_all() for i in index]
+
+    with warnings.catch_warnings():
+        # TODO remove warnings context manager when fixed in upsetplot package
+        warnings.simplefilter(action="ignore", category=FutureWarning)
+        series = upsetplot.from_memberships(
+            labels,
+            data=values,
+        )
+        axes = upsetplot.plot(
+            series, totals_plot_elements=0, with_lines=False, sort_by="input"
+        )
+    axes["intersections"].set_ylabel(r"$\Delta$" + criterion)
+    return axes
+
+
+def line_selected(
+    models: list[Model],
+    criterion: Criterion,
+    relative: bool = True,
+    fz: int = 14,
+    size: tuple[float, float] = (5, 4),
+    labels: dict[str, str] = None,
+    ax: plt.Axes = None,
+) -> plt.Axes:
+    """Plot the improvement in criterion across iterations.
+
+    Args:
+        models:
+            A list of all models. The selected models will be inferred from the
+            best model and its chain of predecessor model hashes.
+        criterion:
+            The criterion by which models are selected.
+        relative:
+            If `True`, criterion values are plotted relative to the lowest
+            criterion value. TODO is the lowest value, always the best? May not
+            be for different criterion.
```

Comment:
Likelihood (`LH`) and log-likelihood (`LLH`) are currently supported (although not really model selection criterion)... This can be handled in a smarter way when when #109 is done, though.

https://github.com/PEtab-dev/petab_select/blob/8d97c58e1f4a671fecb19881827db237f12f51e0/petab_select/model.py#L692-L709